File: desktop/src/app/LogoStage.cpp
#include "LogoStage.hpp"

#include "ModalBackdrop.hpp"   // the modal scrim + blur a showWord wears too
#include "modalReveal.hpp"     // support::motionReduced()

#include <QApplication>
#include <QCursor>
#include <QRandomGenerator>
#include <QTimer>
#include <QToolButton>
#include <cmath>

namespace stencil::gui {

  using support::StageEffect;

  namespace {
    constexpr double PI = 3.14159265358979323846;
    double angle() { return QRandomGenerator::global()->generateDouble() * 2 * PI; }
  }  // namespace

  LogoStage::LogoStage(QWidget* host, QToolButton* logo, Hooks hooks)
      : QWidget(host), hostWindow(host), logo(logo), hooks(std::move(hooks)) {
    setObjectName("logoStage");
    setAttribute(Qt::WA_NoSystemBackground);
    setMouseTracking(true);
    hide();
    hold = new QTimer(this);
    hold->setObjectName("logoHold");
    hold->setSingleShot(true);
    connect(hold, &QTimer::timeout, this, [this] {
      const QString name = heldShow();
      if (name.isEmpty()) return;
      fired = true;
      if (this->hooks.stopClick) this->hooks.stopClick();
      activateByName(name);
    });
    clock = new QTimer(this);
    clock->setTimerType(Qt::PreciseTimer);
    connect(clock, &QTimer::timeout, this, &LogoStage::tick);
    if (this->logo) this->logo->installEventFilter(this);
    qApp->installEventFilter(this);
  }

  int LogoStage::bigEnd(int w, int h) const {
    const bool bouncing = effect == StageEffect::SHRINK || effect == StageEffect::GROW;
    return bouncing ? support::bounceBigSize(w, h) : support::bigLogoSize(w, h);
  }

  QString LogoStage::heldShow() const {
    return support::resolveShow(hooks.accentKey ? hooks.accentKey() : QString(), support::motionMode());
  }

  bool LogoStage::activateByName(const QString& name) {
    const support::StageShow* spec = support::showByName(name);
    if (!spec || open) return false;
    if (hooks.bareWindow && !hooks.bareWindow()) return false;
    // Every showWord's notice is the same: the egg on gold, wearing the golden shining.
    if (spec->effect == StageEffect::PINK) {
      if (hooks.pinkVibe) hooks.pinkVibe();
      if (hooks.toast) hooks.toast(support::logoStageConfig().toast);
      return true;
    }
    start(name);
    if (hooks.toast) hooks.toast(support::logoStageConfig().toast);
    return true;
  }

  void LogoStage::start(const QString& name) {
    const support::StageShow* spec = support::showByName(name);
    showWord = name;
    effect = spec->effect;
    reduced = support::motionReduced();
    takeBackdrop();
    setGeometry(hostWindow->rect());
    const int w = width(), h = height();
    const auto [rest, other] = ends(w, h);
    markPx = rest;
    bounce = support::bounceState(rest, other);
    fly = support::flyState(w / 2.0, h / 2.0, angle());
    chase = support::chaseState(w / 2.0, h / 2.0);
    heading = QPointF(0, 0);
    markCentre = QPointF(w / 2.0, h / 2.0);
    cursorPos = markCentre;
    remakeMark();
    support::ParticleStyle style = support::ParticleStyle::DUST;
    hasCloud = support::showHasCloud(name, &style) && !reduced;
    cloud.setStyle(style, hasCloud);
    // The mark grows out of the header logo, as the browser's does.
    const QPoint at = logo && logo->isVisible() ? logo->mapTo(hostWindow, logo->rect().center()) : rect().center();
    from = support::StagePose{double(at.x()), double(at.y()), double(logo ? logo->iconSize().width() : 32)};
    last = refitAt = 0;
    leftAt = -1;
    open = true;
    held = false;
    boost = 1.0;
    since.restart();
    if (hooks.coverChrome) hooks.coverChrome(true);
    raise();
    show();
    priorFocus = QApplication::focusWidget();
    setFocus(Qt::OtherFocusReason);
    clock->start(support::frameIntervalMs(this));
    update();
  }

  // The mark is measured FROM the hostWindow, so a resize re-measures it and carries everything
  // placed in the old box across as a fraction of it — otherwise the showWord keeps the markPx and the
  // centre of the hostWindow it opened in. Browser twin: the tail of logoStage.js `fit`.
  void LogoStage::relayout() {
    if (!open && leftAt < 0) return;
    const QSize was = size();
    setGeometry(hostWindow->rect());
    const int w = width(), h = height();
    const double kx = was.width() > 0 ? double(w) / was.width() : 1.0;
    const double ky = was.height() > 0 ? double(h) / was.height() : 1.0;
    if (kx == 1.0 && ky == 1.0) return;
    const auto [rest, other] = ends(w, h);
    support::bounceResize(bounce, rest, other);
    markPx = effect == StageEffect::SHRINK || effect == StageEffect::GROW ? bounce.size : rest;
    refitAt = since.elapsed() + REFIT_MS;
    markCentre = QPointF(markCentre.x() * kx, markCentre.y() * ky);
    chase.x *= kx;  chase.y *= ky;
    fly.x *= kx;    fly.y *= ky;
    if (logo && logo->isVisible()) {
      const QPoint at = logo->mapTo(hostWindow, logo->rect().center());
      from = support::StagePose{double(at.x()), double(at.y()), from.size};
    }
    update();
  }

  void LogoStage::dismiss() {
    if (!open) return;
    open = false;
    // Hand the keyboard back, or the hostWindow's own typing — the next word — never reaches it.
    if (priorFocus) priorFocus->setFocus(Qt::OtherFocusReason);
    else if (hostWindow) hostWindow->setFocus(Qt::OtherFocusReason);
    priorFocus.clear();
    if (hooks.coverChrome) hooks.coverChrome(false);
    if (reduced) {
      clock->stop();
      cloud.clear();
      hide();
      return;
    }
    leftAt = since.elapsed();
  }

  void LogoStage::tick() {
    const support::LogoStageConfig& cfg = support::logoStageConfig();
    const double t = since.elapsed();
    const double dt = std::min(50.0, last > 0 ? t - last : 16.7);
    last = t;
    if (refitAt > 0 && t >= refitAt) refit();
    rampBoost(dt);
    if (leftAt >= 0 && t - leftAt >= cfg.hideMs) {
      clock->stop();
      cloud.clear();
      hide();
      leftAt = -1;
      return;
    }
    if (!reduced && leftAt < 0) {
      const int w = width(), h = height();
      // The pointer is SAMPLED each frame, never taken on trust from a move event: one that does
      // not reach the stage leaves the chase with nothing to chase and the hand never applied.
      const QPoint here = mapFromGlobal(QCursor::pos());
      if (rect().contains(here)) cursorPos = QPointF(here);
      if (effect == StageEffect::FOLLOW || effect == StageEffect::ESCAPE) {
        support::chaseStep(chase, cursorPos, dt, markPx, w, h, effect == StageEffect::ESCAPE);
        markCentre = QPointF(chase.x, chase.y);
        heading = support::headingOfState(chase.vx, chase.vy);
      } else if (effect == StageEffect::FLY) {
        support::flyStep(fly, dt, markPx, w, h);
        markCentre = QPointF(fly.x, fly.y);
        // No tail: this one is not chasing anything, so its cloud stays a ring on every side.
        heading = QPointF(0, 0);
      }
      if (effect == StageEffect::SHRINK || effect == StageEffect::GROW) markPx = support::bounceStep(bounce, t);
      // Spawning tapers with the fade, so the hide has nothing new arriving into it.
      const double fade = leftAt < 0 ? std::min(1.0, t / double(cfg.revealMs))
                                      : std::max(0.0, 1.0 - (t - leftAt) / double(cfg.hideMs));
      cloud.step(dt, markPx, boostNow() * fade, heading);
      syncCursor();
    }
    update();
  }

  // The catch is the mark's OWN outline, not a circle round it: a circle reaches past the flat
  // edges and falls short of the corners, so the hand showed where the art was not. A fast mark
  // is past the pointer by the time a press lands, so it reaches ahead by grabLeadMs of travel.
  bool LogoStage::onMark(const QPoint& at) const {
    const double dx = at.x() - markCentre.x(), dy = at.y() - markCentre.y();
    const bool chasing = effect == StageEffect::FOLLOW || effect == StageEffect::ESCAPE;
    const double v = effect == StageEffect::FLY ? std::hypot(fly.vx, fly.vy)
                   : chasing ? std::hypot(chase.vx, chase.vy) : 0.0;
    const QPointF edge = support::markEdge(markPx, std::atan2(dy, dx));
    const double lead = (v * support::logoStageConfig().flyGrabLeadMs) / 1000.0;
    return std::hypot(dx, dy) <= std::hypot(edge.x(), edge.y()) + lead;
  }

  // ONLY a press drives the light and the cloud: crossing the mark changes nothing.
  double LogoStage::boostNow() const { return boost; }

  // The press EASES in and out over its own time constant — stepped to the target, the light
  // and the grain count both jumped.
  void LogoStage::rampBoost(double dt) {
    const support::LogoStageConfig& cfg = support::logoStageConfig();
    const double target = held ? cfg.holdBoost : 1.0;
    const double k = reduced ? 1.0 : 1.0 - std::exp(-dt / double(cfg.holdRampMs));
    boost += (target - boost) * k;
  }

  // Off the mark, a press closes the stage. On it, a press is a HOLD: the light and the cloud
  // swell for as long as it lasts, and only the shows with a move of their own act on it.
  void LogoStage::pressed(const QPoint& at) {
    if (!onMark(at)) { dismiss(); return; }
    held = true;
    if (effect == StageEffect::SHRINK || effect == StageEffect::GROW)
      support::bounceImpulse(bounce, since.elapsed());
    else if (effect == StageEffect::FLY)
      support::flyPunch(fly, angle());
  }

}  // namespace stencil::gui

File: browser/js/ui/logo/logoStage.js
// The logo stage: one full-window canvas holding the big mark, its light and its cloud, and the
// lock that makes it the only thing the editor listens to until Escape or a click ends it.
// Desktop twin: app/LogoStage.{hpp,cpp}.
import { motionReduced, particleStyle } from '../motionPrefs.js';
import { modalShells } from '../modal/modalRegistry.js';
import { STAGE, effectOf, showStyle, bigLogoSize, bounceBigSize, minLogoSize, roams, roamLogoSize, markEdge } from './logoStageRules.js';
import { beatAt, spinAt, stageMark, markHex, paintBackdrop, paintGlow, paintSpokes, paintMark } from './logoStagePaint.js';
import { createStageCloud } from './logoStageCloud.js';
import * as M from './logoStageMotion.js';

export const STAGE_CLASS = 'logo-stage';
export const OPEN_CLASS = 'logo-stage-open';
let live = null;

export const logoStageOpen = () => !!live;

// Only from the bare window: no stage already up, not fullscreen, nothing modal on top.
export const logoStageAllowed = (doc = globalThis.document) => {
  if (live || typeof doc?.body?.appendChild !== 'function' || typeof doc.createElement !== 'function') return false;
  if (doc.body.classList?.contains('fullscreen-mode')) return false;
  for (const shell of modalShells) if (shell.isOpen()) return false;
  return !doc.querySelector?.('.app-modal-overlay.modal-open');
};

const SWALLOWED = ['keydown', 'keyup', 'keypress', 'wheel', 'contextmenu'];

export const closeLogoStage = () => {
  if (!live) return false;
  const stage = live;
  live = null;
  stage.release();
  return true;
};

// Where the header mark sits, so a show opened by a typed word or a facade call grows out of
// the logo exactly as a held one does.
export const markOrigin = (doc = globalThis.document) => {
  const r = doc?.querySelector?.('.app-logo-wrap')?.getBoundingClientRect?.();
  return r && (r.width || r.height) ? { x: r.left + r.width / 2, y: r.top + r.height / 2 } : null;
};

// `origin` is the header mark's centre, so the stage grows out of the logo and shrinks back into it.
export const openLogoStage = (name, { app, origin = null, doc = globalThis.document } = {}) => {
  const effect = effectOf(name);
  if (!effect || effect === 'pink' || !logoStageAllowed(doc)) return false;
  const reduced = motionReduced();
  const host = doc.createElement('div');
  host.className = STAGE_CLASS;
  host.setAttribute('aria-hidden', 'true');
  const canvas = doc.createElement('canvas');
  host.appendChild(canvas);
  doc.body.appendChild(host);
  // The show's own notice is the one thing that still stands above it (components/logoStage.css).
  doc.body.classList?.add?.(OPEN_CLASS);
  const ctx = canvas.getContext?.('2d') || null;
  const img = stageMark(doc, markHex(app));
  const style = reduced ? null : showStyle(name, particleStyle());
  const cloud = createStageCloud(style);

  const view = () => ({ w: globalThis.innerWidth || 1024, h: globalThis.innerHeight || 768 });
  let { w, h } = view();
  const roaming = roams(name);
  const bouncing = effect === 'shrink' || effect === 'grow';
  const big = () => (bouncing ? bounceBigSize(w, h) : bigLogoSize(w, h));
  const measure = () => (roaming ? roamLogoSize(w, h) : effect === 'grow' ? minLogoSize(w, h) : big());
  const measureOther = () => (effect === 'grow' ? big() : minLogoSize(w, h));
  let rest = measure(), other = measureOther();
  const bounce = M.bounceState(rest, other);
  const fly = M.flyState(w / 2, h / 2, Math.random() * 2 * Math.PI);
  const chase = M.chaseState(w / 2, h / 2);
  const pos = { x: w / 2, y: h / 2 };
  const cursor = { x: w / 2, y: h / 2 };
  let heading = null;   // the way it travels; the cloud lays its tail the other way
  const at = origin || markOrigin(doc);
  const from = at ? { x: at.x, y: at.y, size: 32 } : { x: w / 2, y: h / 2, size: rest };
  let size = rest, held = false, leftAt = null;
  // ONLY a press drives the light and the cloud: crossing the mark changes nothing, so the
  // loop never restarts under a passing pointer. The press EASES it in and out over its own
  // time constant — stepped to the target, the light and the grain count both jumped.
  let boost = 1;
  const rampBoost = (dt) => {
    const target = held ? STAGE.stage.holdBoost : 1;
    const k = reduced ? 1 : 1 - Math.exp(-dt / STAGE.stage.holdRampMs);
    boost += (target - boost) * k;
  };
  let t0 = 0, last = 0, frame = null, fade = reduced ? 1 : 0;

  const fit = () => {
    const was = { w, h };
    ({ w, h } = view());
    const dpr = Math.min(globalThis.devicePixelRatio || 1, 2);
    canvas.width = Math.round(w * dpr);
    canvas.height = Math.round(h * dpr);
    canvas.style.width = `${w}px`;
    canvas.style.height = `${h}px`;
    ctx?.setTransform(dpr, 0, 0, dpr, 0, 0);
    // The mark is measured FROM the window, so a resize re-measures it and carries everything
    // placed in the old box across as a fraction of it — otherwise the show keeps the size and
    // the centre of the window it opened in.
    const kx = was.w > 0 ? w / was.w : 1, ky = was.h > 0 ? h / was.h : 1;
    if (kx === 1 && ky === 1) return;
    rest = measure();
    other = measureOther();
    M.bounceResize(bounce, rest, other);
    size = effect === 'shrink' || effect === 'grow' ? bounce.size : rest;
    pos.x *= kx; pos.y *= ky;
    chase.x *= kx; chase.y *= ky;
    fly.x *= kx; fly.y *= ky;
  };
  fit();

  const drop = () => { host.remove(); doc.body.classList?.remove?.(OPEN_CLASS); };
  const now = () => (globalThis.performance?.now?.() ?? Date.now());
  const paint = (t) => {
    if (!ctx) return;
    const beat = reduced ? 0.5 : beatAt(t);
    // A cloud show's light is a steady lamp the grains fly through: all its motion, and
    // everything a hold adds, belongs to the cloud. Without a cloud the light does it all.
    const lightBoost = style ? 1 : boost;
    // Never all the way down: a neon sign breathes, it does not go out.
    const lit = style ? STAGE.glow.steadyLit : STAGE.glow.floor + (1 - STAGE.glow.floor) * beat;
    const reveal = reduced ? 1 : Math.min(1, t / STAGE.stage.revealMs);
    const p = leftAt === null ? reveal : 1 - Math.min(1, (t - leftAt) / STAGE.stage.hideMs);
    const pose = M.revealTween(from, { x: pos.x, y: pos.y, size }, p);
    paintBackdrop(ctx, w, h, fade);
    // The cloud wears the mark's own scale, so it grows out of the logo and shrinks back into it.
    cloud.draw(ctx, doc, pose.x, pose.y, t, size > 0 ? pose.size / size : 1);
    paintGlow(ctx, { ...pose, hex: markHex(app), beat: lit, boost: lightBoost });
    if (effect === 'sun') paintSpokes(ctx, { ...pose, hex: markHex(app), beat, boost: lightBoost, angle: reduced ? 0 : spinAt(t) });
    paintMark(ctx, img, pose);
  };

  const step = () => {
    frame = null;
    const t = now() - t0;
    const dt = Math.min(50, last ? t - last : 16.7);
    last = t;
    rampBoost(dt);
    fade = leftAt === null ? Math.min(1, t / STAGE.stage.revealMs)
      : Math.max(0, 1 - (t - leftAt) / STAGE.stage.hideMs);
    if (leftAt !== null && t - leftAt >= STAGE.stage.hideMs) { drop(); return; }
    if (!reduced) {
      if (effect === 'follow' || effect === 'escape') {
        M.chaseStep(chase, cursor, dt, size, w, h, STAGE[effect], effect === 'escape');
        pos.x = chase.x; pos.y = chase.y;
        heading = M.headingOfState(chase);
      } else if (effect === 'fly') {
        M.flyStep(fly, dt, size, w, h);
        pos.x = fly.x; pos.y = fly.y;
        // No tail: this one is not chasing anything, so its cloud stays a ring on every side.
        heading = null;
      }
      size = effect === 'shrink' || effect === 'grow' ? M.bounceStep(bounce, t) : size;
      // Spawning tapers with the fade, so the hide has nothing new arriving into it.
      cloud.step(dt, size, boost * fade, { dir: heading });
      syncCursor();
    }
    paint(t);
    frame = raf(step);
  };
  const raf = globalThis.requestAnimationFrame ?? ((fn) => setTimeout(() => fn(now()), 16));
  const cancel = globalThis.cancelAnimationFrame ?? clearTimeout;

  // The catch is the mark's OWN outline, not a circle round it: a circle reaches past the flat
  // edges and falls short of the corners, so the hand showed where the art was not. A fast mark
  // is past the pointer by the time a press lands, so it reaches ahead by grabLeadMs of travel.
  const onMark = (x, y) => {
    const dx = x - pos.x, dy = y - pos.y;
    const v = effect === 'fly' ? Math.hypot(fly.vx, fly.vy)
            : effect === 'follow' || effect === 'escape' ? Math.hypot(chase.vx, chase.vy) : 0;
    const edge = markEdge(size, Math.atan2(dy, dx));
    return Math.hypot(dx, dy) <= Math.hypot(edge.x, edge.y) + (v * STAGE.fly.grabLeadMs) / 1000;
  };
  // The mark is the one thing on the stage a press acts on, so it wears the hand — and a roaming
  // mark runs under a still pointer, so the frame re-asks as well as the move.
  let hand = null;
  const syncCursor = () => {
    const hot = onMark(cursor.x, cursor.y);
    if (hot !== hand) host.style.cursor = (hand = hot) ? 'pointer' : 'default';
  };
  const onPointerMove = (e) => {
    cursor.x = e.clientX ?? cursor.x;
    cursor.y = e.clientY ?? cursor.y;
    syncCursor();
  };
  // Off the mark, a press closes the stage. On it, a press is a HOLD: the light and the cloud
  // swell for as long as it lasts, and only the shows with a move of their own act on it.
  const onPointerDown = (e) => {
    e.preventDefault?.();
    if (!onMark(e.clientX ?? 0, e.clientY ?? 0)) { closeLogoStage(); return; }
    held = true;
    if (effect === 'shrink' || effect === 'grow') M.bounceImpulse(bounce, now() - t0);
    else if (effect === 'fly') M.flyPunch(fly, Math.random() * 2 * Math.PI);
  };
  const onPointerUp = () => { held = false; };
  // Capture, so nothing the editor bound ever sees a key while the stage is up.
  const swallow = (e) => {
    if (e.type === 'keydown' && e.key === 'Escape') { closeLogoStage(); return; }
    e.preventDefault?.();
    e.stopImmediatePropagation?.();
  };

  host.addEventListener('pointermove', onPointerMove);
  host.addEventListener('pointerdown', onPointerDown);
  for (const type of ['pointerup', 'pointercancel', 'pointerleave']) host.addEventListener(type, onPointerUp);
  for (const type of SWALLOWED) doc.addEventListener(type, swallow, true);
  globalThis.addEventListener?.('resize', fit);

  live = {
    name, effect, host, canvas,
    get size() { return size; },
    get position() { return { ...pos }; },
    get cloudLive() { return cloud.live; },
    release() {
      leftAt = now() - t0;
      host.removeEventListener('pointermove', onPointerMove);
      host.removeEventListener('pointerdown', onPointerDown);
      for (const type of ['pointerup', 'pointercancel', 'pointerleave']) host.removeEventListener(type, onPointerUp);
      for (const type of SWALLOWED) doc.removeEventListener(type, swallow, true);
      globalThis.removeEventListener?.('resize', fit);
      if (reduced) { if (frame) cancel(frame); drop(); }
    },
  };
  t0 = now();
  step();
  return true;
};

// Tests and the facade: the stage now on screen, or null.
export const currentLogoStage = () => live;
